Document Page Builder initial support (#264)
front-commerce/developers.front-commerce.com: source/docs/magento2/page-builder.md

---
id: magento2-page-builder
title: Page Builder
---

Adobe Commerce and Magento 2.4.3+ allow merchants to author pages using [Page Builder](https://magento.com/products/magento-commerce/page-builder). Front-Commerce supports Page Builder managed content out-of-the-box.

In this section, you will learn how to use this feature and extend it. Create new *content types* and refine UI components so that merchants can create the rich shopping experiences that was designed for *their* customers.

<blockquote class="feature--new">
_Since version 2.11.0 (early preview)_
<br>
**Early preview:** we're currently looking for feedbacks about this feature. The core API is stable and ready for use in your storefronts. We're improving default content types support. Please contact us if you're using this feature.
<br>

> **WIP:** this documentation page will get more details in the next few days

Reviewer: ```suggestion
**Work In Progress:** this documentation page will get more details in the next few days
```

</blockquote>

- override `theme/modules/WysiwygV2/MagentoWysiwyg/PageBuilder/_appComponents.scss` to register your custom styles
- override `theme/modules/WysiwygV2/MagentoWysiwyg/PageBuilder/appComponentsMap.js` to register new components (or override [existing ones](https://gitlab.com/front-commerce/front-commerce/blob/main/src/web/theme/modules/WysiwygV2/MagentoWysiwyg/PageBuilder/index.js))

<!-- Override GraphQL fragment too (not yet externalized in a specific fragment FC code) -->

### Expose content types data in GraphQL

<blockquote class="wip">
**Work In Progress:** if you need details right now, please [contact us](mailto:contact@front-commerce.com). We will make sure to answer you in a timely manner.
</blockquote>

The `PageBuilder` loader allows you to register new content types.

First, you must define your content type. Content types must extend the `ContentType` class (see below). The `name` should match Magento's content type identifier and the `extractData` method can be used to return structured data to be exposed as GraphQL fields for this type.

> ```

Reviewer: ```suggestion
```javascript
```

import ContentType from "server/modules/magento2/wysiwyg/loaders/content-types/ContentType";

export default class Foo extends ContentType {

>   name = "foo";

Reviewer: it would be nice to find a realistic (but not too complicated example)

(can be made in a follow up as the page is WIP anyway)

Author: Yes, that's a good idea! I'll update it later.

  extractData(node) {
    return {
      // custom data. Can also be extracted from the passed `node` information
      bar: "baz"
    };
  }
}
```

Then you must register it from a `contextEnhancer`, using the `PageBuilder.registerContentType` method:

> ```

Reviewer: ```suggestion
```javascript
```
